R/ensure-source.R

```diff
@@ -180,17 +148,98 @@ ensure_dataset <- function(name, download = TRUE) {
-      arrow::open_dataset(path, partitioning = c("year", "month"))
-    },
-    dim = c(1547741381L, 20L),
-    n_files = 125
-  )
-)
+source_filename <- function(name) {
+  filename <- get_source_attr(name, "url")
+
+  # if the filename is NULL, this is a test data source
+  if (is.null(filename)) {
+    filename <- get_source_attr(name, "filename")
+  }
+
+  ext <- file_ext(basename(filename))
+  paste(c(name, ext), collapse = ".")
+}
+
+#' Ensure that a source has a specific format
+#'
+#' @param name name of the known source
+#' @param format format to be ensured
+#' @param compression compression to be ensured
+#'
+#' @return the file that was ensured to exist
+#' @export
+#'
+#' @importFrom utils write.csv
+ensure_format <- function(
+  name,
+  format = c("csv", "parquet", "feather", "csv.gz"),
+  compression = c("uncompressed", "snappy", "zstd", "gzip", "gz", "lz4")) {
+  compression <- match.arg(compression)
+  format <- match.arg(format)
+
+  # if we get "csv.gz" split it up correctly.
+  if (format == "csv.gz") {
+    format <- "csv"
+    compression <- "gzip"
+  }
+
+  # generate an extension based on format + compression (with special cases for
+  # uncompressed and gzip)
+  if (compression == "gzip") {
+    ext <- paste(format, "gz", sep = ".")
+  } else if (compression == "uncompressed") {
+    ext <- format
+  } else {
+    ext <- paste(format, compression, sep = ".")
+  }
+
+  # exit quickly if exists already
+  file_out <- data_file(file_with_ext(source_filename(name), ext))
+  if (file.exists(file_out)) {
+    return(file_out)
+  }
+
+  # special case if input is csv + gzip compression since we don't need to read
+  # that just to compress
+  file_in <- ensure_source(name)
+  if(format == "csv" & file_ext(file_in) == "csv" ) {
+    if(compression == "gzip" & file_ext(file_in) == "csv") {
+      # compress if the file doesn't already exist
+      R.utils::gzip(file_in, file_out, remove = FALSE)
+      return(file_out)
+    } else if(compression == "uncompressed" & file_ext(file_in) == "csv.gz") {
+      # compress if the file doesn't already exist
+      R.utils::gunzip(file_in, file_out, remove = FALSE)
+      return(file_out)
+    }
+    return(file_in)
+  }
+
+  # read the data in
+  # TODO: read in things that are easier to read in feather > parquet >> csv?
+  source <- all_sources[[name]]
+  tab <- source$reader(ensure_source(name))
+
+  # write the reformatted data based on the format/ext
+  if (format == "csv") {
+    csv_filename <- gsub(".gz", "", file_out)
+    write.csv(tab, file = csv_filename, row.names = FALSE, na = "")
+
+    if (compression == "gzip") {
+      R.utils::gzip(csv_filename, ext = "gz", remove = FALSE)
+    }
+  } else if (format == "parquet") {
+    if (compression == "gzip") {
+      stop("gzip + parquet is not a known compression", call. = FALSE)
+    }
+    arrow::write_parquet(tab, sink = file_out, compression = compression)
+  } else if (format == "feather") {
+    if (compression == "gzip") {
```

Comment:
You also can't write feather + snappy. 

We seem to have this logic also in the valid_params methods for read_file and write_file benchmarks, so maybe we need a centralized function/object that tells you what file formats and compressions are valid combinations.

Reply:
Yeah, I like the idea of keeping the valid options centrally. I hadn't (yet) pulled out the code in `read_file` and `write_file` that those use to do the transformation, but that was my plan for next steps was to pull that out and use the new `ensure_format()` and it would be good to use a single place to figure out what options are possible here and in the benchmarks.

FWIW, feather + snappy should never be triggered since the `write_file` and `read_file` should never ask for that, but I agree centralization for this is a better solution.